Tests/SentryTests/Integrations/SentryFramesTrackerTests.swift

```diff
@@ -0,0 +1,131 @@
+import XCTest
+
+#if os(iOS) || os(tvOS) || targetEnvironment(macCatalyst)
+class SentryFramesTrackerTests: XCTestCase {
+    
+    private class Fixture {
+        
+        var displayLinkWrapper: TestDiplayLinkWrapper
+        var queue: DispatchQueue
+        
+        init() {
+            displayLinkWrapper = TestDiplayLinkWrapper()
+            queue = DispatchQueue(label: "SentryFramesTrackerTests", qos: .background, attributes: [.concurrent])
+        }
+        
+        var sut: SentryFramesTracker {
+            return SentryFramesTracker(displayLinkWrapper: displayLinkWrapper)
+        }
+    }
+    
+    private var fixture: Fixture!
+    
+    override func setUp() {
+        fixture = Fixture()
+    }
+    
+    func testIsNotRunning_WhenNotStarted() {
+        XCTAssertFalse(fixture.sut.isRunning)
+    }
+    
+    func testIsRunning_WhenStarted() {
+        let sut = fixture.sut
+        sut.start()
+        XCTAssertTrue(sut.isRunning)
+    }
+    
+    func testIsNotRunning_WhenStopped() {
+        let sut = fixture.sut
+        sut.start()
+        sut.stop()
+        XCTAssertFalse(sut.isRunning)
+    }
+    
+    func testSlowFrame() {
+        let sut = fixture.sut
+        sut.start()
+        
+        fixture.displayLinkWrapper.call()
+        fixture.displayLinkWrapper.internalTimestamp += TestData.slowFrameThreshold + 0.001
+        fixture.displayLinkWrapper.call()
+        fixture.displayLinkWrapper.internalTimestamp += TestData.slowFrameThreshold
+        fixture.displayLinkWrapper.call()
+        fixture.displayLinkWrapper.internalTimestamp += TestData.frozenFrameThreshold
+        fixture.displayLinkWrapper.call()
+        
+        XCTAssertEqual(2, sut.currentSlowFrames)
+        XCTAssertEqual(3, sut.currentTotalFrames)
+        XCTAssertEqual(0, sut.currentFrozenFrames)
+    }
+    
+    func testFrozenFrame() {
+        let sut = fixture.sut
+        sut.start()
+        
+        fixture.displayLinkWrapper.call()
+        fixture.displayLinkWrapper.internalTimestamp += TestData.frozenFrameThreshold + 0.001
+        fixture.displayLinkWrapper.call()
+        fixture.displayLinkWrapper.internalTimestamp += TestData.frozenFrameThreshold
+        fixture.displayLinkWrapper.call()
+        
+        XCTAssertEqual(1, sut.currentSlowFrames)
+        XCTAssertEqual(2, sut.currentTotalFrames)
+        XCTAssertEqual(1, sut.currentFrozenFrames)
+    }
+    
+    func testAllFrames_ConcurrentRead() {
+        let sut = fixture.sut
+        sut.start()
+        
+        let group = DispatchGroup()
+        
+        assertPreviousCountBiggerThanCurrent(group) { return sut.currentFrozenFrames }
+        assertPreviousCountBiggerThanCurrent(group) { return sut.currentSlowFrames }
+        assertPreviousCountBiggerThanCurrent(group) { return sut.currentTotalFrames }
+        
+        fixture.displayLinkWrapper.call()
+        
+        let frames: UInt = 600_000
```

Comment:
l: how fast does this run? wondering if it makes sense to slow down the test, a lower number would result in the same way if deterministic, I don't think 600k is that much though

Reply:
It takes around 1 to 2 seconds. I rather keep it that big because I had synchronization issues in the past, and with a lower number, I had to run the test a few times to uncover them.